### DISCUSSION TITLE
Use pointmatcher with CGAL fail

### DISCUSSION THREAD

**Discussion by linbeijianbaoxia** [CGAL/cgal#8123] [Q&A]
I tried to run the PointMatcher example program from CGAL's Point Set Processing: https://doc.cgal.org/latest/Point_set_processing_3/index.html#Point_set_processing_3Examples_registration_PointMatcher. I have successfully installed libpointmatcher, version 1.3.1. When I tried to compile the example project using CMake, `find_package(libpointmatcher QUIET)` also successfully found my libpointmatcher library, but it seems that `if(TARGET CGAL::pointmatcher_support)` did not pass, resulting in the corresponding project not being compiled. This has confused me about where the problem lies. Below is the information output during CMake compilation:

```
Selecting Windows SDK version 10.0.19041.0 to target Windows 10.0.19045.
Visual Leak Detector (VLD) is not found.
Using header-only CGAL
Targeting Visual Studio 16 2019
Target build environment supports auto-linking
Using VC toolset 142.
Generator uses intermediate configuration directory: $(Configuration)
Boost include dirs: C:/dev/vcpkg/installed/x64-windows/include
Boost libraries:    
USING DEBUG CXXFLAGS   = '/DWIN32 /D_WINDOWS /GR /EHsc /Zi /Ob0 /Od /RTC1'
USING DEBUG EXEFLAGS   = '/machine:x64 /debug /INCREMENTAL'
USING RELEASE CXXFLAGS = '/DWIN32 /D_WINDOWS /GR /EHsc /O2 /Ob2 /DNDEBUG'
USING RELEASE EXEFLAGS = '/machine:x64 /INCREMENTAL:NO'
NOTICE: the LAS reader example requires LASlib and will not be compiled.
CMake Warning at D:/Program Files/CMake/share/cmake-3.19/Modules/FindBoost.cmake:1204 (message):
  New Boost version may have incorrect or missing dependencies and imported
  targets
Call Stack (most recent call first):
  D:/Program Files/CMake/share/cmake-3.19/Modules/FindBoost.cmake:1326 (_Boost_COMPONENT_DEPENDENCIES)
  D:/Program Files/CMake/share/cmake-3.19/Modules/FindBoost.cmake:1935 (_Boost_MISSING_DEPENDENCIES)
  C:/dev/vcpkg/installed/x64-windows/share/boost/vcpkg-cmake-wrapper.cmake:11 (_find_package)
  C:/dev/vcpkg/scripts/buildsystems/vcpkg.cmake:813 (include)
  C:/dev/vcpkg/installed/x64-windows/share/cgal/CGAL_pointmatcher_support.cmake:2 (find_package)
  CMakeLists.txt:96 (include)


CMake Warning at D:/Program Files/CMake/share/cmake-3.19/Modules/FindBoost.cmake:1204 (message):
  New Boost version may have incorrect or missing dependencies and imported
  targets
Call Stack (most recent call first):
  D:/Program Files/CMake/share/cmake-3.19/Modules/FindBoost.cmake:1326 (_Boost_COMPONENT_DEPENDENCIES)
  D:/Program Files/CMake/share/cmake-3.19/Modules/FindBoost.cmake:1935 (_Boost_MISSING_DEPENDENCIES)
  C:/dev/vcpkg/installed/x64-windows/share/boost/vcpkg-cmake-wrapper.cmake:11 (_find_package)
  C:/dev/vcpkg/scripts/buildsystems/vcpkg.cmake:813 (include)
  C:/dev/vcpkg/installed/x64-windows/share/cgal/CGAL_pointmatcher_support.cmake:2 (find_package)
  CMakeLists.txt:96 (include)


CMake Warning at D:/Program Files/CMake/share/cmake-3.19/Modules/FindBoost.cmake:1204 (message):
  New Boost version may have incorrect or missing dependencies and imported
  targets
Call Stack (most recent call first):
  D:/Program Files/CMake/share/cmake-3.19/Modules/FindBoost.cmake:1326 (_Boost_COMPONENT_DEPENDENCIES)
  D:/Program Files/CMake/share/cmake-3.19/Modules/FindBoost.cmake:1935 (_Boost_MISSING_DEPENDENCIES)
  C:/dev/vcpkg/installed/x64-windows/share/boost/vcpkg-cmake-wrapper.cmake:11 (_find_package)
  C:/dev/vcpkg/scripts/buildsystems/vcpkg.cmake:813 (include)
  C:/dev/vcpkg/installed/x64-windows/share/cgal/CGAL_pointmatcher_support.cmake:2 (find_package)
  CMakeLists.txt:96 (include)


CMake Warning at D:/Program Files/CMake/share/cmake-3.19/Modules/FindBoost.cmake:1204 (message):
  New Boost version may have incorrect or missing dependencies and imported
  targets
Call Stack (most recent call first):
  D:/Program Files/CMake/share/cmake-3.19/Modules/FindBoost.cmake:1326 (_Boost_COMPONENT_DEPENDENCIES)
  D:/Program Files/CMake/share/cmake-3.19/Modules/FindBoost.cmake:1935 (_Boost_MISSING_DEPENDENCIES)
  C:/dev/vcpkg/installed/x64-windows/share/boost/vcpkg-cmake-wrapper.cmake:11 (_find_package)
  C:/dev/vcpkg/scripts/buildsystems/vcpkg.cmake:813 (include)
  C:/dev/vcpkg/installed/x64-windows/share/cgal/CGAL_pointmatcher_support.cmake:2 (find_package)
  CMakeLists.txt:96 (include)


CMake Warning at D:/Program Files/CMake/share/cmake-3.19/Modules/FindBoost.cmake:1204 (message):
  New Boost version may have incorrect or missing dependencies and imported
  targets
Call Stack (most recent call first):
  D:/Program Files/CMake/share/cmake-3.19/Modules/FindBoost.cmake:1326 (_Boost_COMPONENT_DEPENDENCIES)
  D:/Program Files/CMake/share/cmake-3.19/Modules/FindBoost.cmake:1935 (_Boost_MISSING_DEPENDENCIES)
  C:/dev/vcpkg/installed/x64-windows/share/boost/vcpkg-cmake-wrapper.cmake:11 (_find_package)
  C:/dev/vcpkg/scripts/buildsystems/vcpkg.cmake:813 (include)
  C:/dev/vcpkg/installed/x64-windows/share/cgal/CGAL_pointmatcher_support.cmake:2 (find_package)
  CMakeLists.txt:96 (include)


CMake Warning at D:/Program Files/CMake/share/cmake-3.19/Modules/FindBoost.cmake:1204 (message):
  New Boost version may have incorrect or missing dependencies and imported
  targets
Call Stack (most recent call first):
  D:/Program Files/CMake/share/cmake-3.19/Modules/FindBoost.cmake:1326 (_Boost_COMPONENT_DEPENDENCIES)
  D:/Program Files/CMake/share/cmake-3.19/Modules/FindBoost.cmake:1935 (_Boost_MISSING_DEPENDENCIES)
  C:/dev/vcpkg/installed/x64-windows/share/boost/vcpkg-cmake-wrapper.cmake:11 (_find_package)
  C:/dev/vcpkg/scripts/buildsystems/vcpkg.cmake:813 (include)
  C:/dev/vcpkg/installed/x64-windows/share/cgal/CGAL_pointmatcher_support.cmake:2 (find_package)
  CMakeLists.txt:96 (include)


CMake Warning at D:/Program Files/CMake/share/cmake-3.19/Modules/FindBoost.cmake:1204 (message):
  New Boost version may have incorrect or missing dependencies and imported
  targets
Call Stack (most recent call first):
  D:/Program Files/CMake/share/cmake-3.19/Modules/FindBoost.cmake:1326 (_Boost_COMPONENT_DEPENDENCIES)
  D:/Program Files/CMake/share/cmake-3.19/Modules/FindBoost.cmake:1935 (_Boost_MISSING_DEPENDENCIES)
  C:/dev/vcpkg/installed/x64-windows/share/boost/vcpkg-cmake-wrapper.cmake:11 (_find_package)
  C:/dev/vcpkg/scripts/buildsystems/vcpkg.cmake:813 (include)
  C:/dev/vcpkg/installed/x64-windows/share/cgal/CGAL_pointmatcher_support.cmake:2 (find_package)
  CMakeLists.txt:96 (include)


Could NOT find Boost (missing: program_options) (found version "1.83.0")
NOTICE: the libpointmatcher library requires the following boost components: thread filesystem system program_options date_time chrono.
NOTICE: registration with pointmatcher requires libpointmatcher and will not be compiled.
NOTICE: registration with OpenGR and pointmatcher requires both libpointmatcher and OpenGR, and will not be compiled.
Configuring done
```

I'm not sure if it's related to the Boost version, as my CGAL was installed directly using vcpkg.
![image](https://github.com/CGAL/cgal/assets/50166977/d367aa43-2560-4ed0-9b26-1c26a7577122)

## Environment

* Operating system (Windows/Mac/Linux, 32/64 bits):windows 10 x64
* Compiler:MSVC 2019
* Release or debug mode:
* Specific flags used (if any): 
* CGAL version:5.6.1
* Boost version:1.83.0
* Other libraries versions if used (Eigen, TBB, etc.):OpenGR PointMatcher

**Comment by lrineau** [comment]
`C:/dev/vcpkg/installed/x64-windows/share/cgal/CGAL_pointmatcher_support.cmake` searchs for dependencies of libpointmatcher. And we see in your log that it displays:

```
NOTICE: the libpointmatcher library requires the following boost components: thread filesystem system program_options date_time chrono.
```

Part of Boost libraries are not found correctly.

**Reply by linbeijianbaoxia** [reply]
According to the CMake prompt, I installed "boost-preprocessor:x64-windows 1.83.0 Boost preprocessor module" using vcpkg, and then proceeded with the configuration again. However, an error occurred during the generation process. The CMake output is as follows:“Selecting Windows SDK version 10.0.19041.0 to target Windows 10.0.19045.
The C compiler identification is MSVC 19.16.27045.0
The CXX compiler identification is MSVC 19.16.27045.0
Detecting C compiler ABI info
Detecting C compiler ABI info - done
Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2017/Community/VC/Tools/MSVC/14.16.27023/bin/Hostx86/x64/cl.exe - skipped
Detecting C compile features
Detecting C compile features - done
Detecting CXX compiler ABI info
Detecting CXX compiler ABI info - done
Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2017/Community/VC/Tools/MSVC/14.16.27023/bin/Hostx86/x64/cl.exe - skipped
Detecting CXX compile features
Detecting CXX compile features - done
Visual Leak Detector (VLD) is not found.
Using header-only CGAL
Targeting Visual Studio 15 2017
Target build environment supports auto-linking
Using VC toolset 141.
Generator uses intermediate configuration directory: $(Configuration)
Found GMP: debug;C:/dev/vcpkg/installed/x64-windows/debug/lib/gmp.lib;optimized;C:/dev/vcpkg/installed/x64-windows/debug/lib/gmp.lib  
Found MPFR: C:/dev/vcpkg/installed/x64-windows/debug/lib/mpfr.lib  
Found Boost: C:/dev/vcpkg/installed/x64-windows/include (found suitable version "1.83.0", minimum required is "1.66")  
Boost include dirs: C:/dev/vcpkg/installed/x64-windows/include
Boost libraries:    
Looking for pthread.h
Looking for pthread.h - not found
Found Threads: TRUE  
USING DEBUG CXXFLAGS   = '/DWIN32 /D_WINDOWS /GR /EHsc /Zi /Ob0 /Od /RTC1'
USING DEBUG EXEFLAGS   = '/machine:x64 /debug /INCREMENTAL'
USING RELEASE CXXFLAGS = '/DWIN32 /D_WINDOWS /GR /EHsc /O2 /Ob2 /DNDEBUG'
USING RELEASE EXEFLAGS = '/machine:x64 /INCREMENTAL:NO'
NOTICE: the LAS reader example requires LASlib and will not be compiled.
CMake Warning at D:/Program Files/CMake/share/cmake-3.19/Modules/FindBoost.cmake:1204 (message):
  New Boost version may have incorrect or missing dependencies and imported
  targets
Call Stack (most recent call first):
  D:/Program Files/CMake/share/cmake-3.19/Modules/FindBoost.cmake:1326 (_Boost_COMPONENT_DEPENDENCIES)
  D:/Program Files/CMake/share/cmake-3.19/Modules/FindBoost.cmake:1935 (_Boost_MISSING_DEPENDENCIES)
  C:/dev/vcpkg/installed/x64-windows/share/boost/vcpkg-cmake-wrapper.cmake:11 (_find_package)
  C:/dev/vcpkg/scripts/buildsystems/vcpkg.cmake:813 (include)
  C:/dev/vcpkg/installed/x64-windows/share/cgal/CGAL_pointmatcher_support.cmake:2 (find_package)
  CMakeLists.txt:96 (include)


CMake Warning at D:/Program Files/CMake/share/cmake-3.19/Modules/FindBoost.cmake:1204 (message):
  New Boost version may have incorrect or missing dependencies and imported
  targets
Call Stack (most recent call first):
  D:/Program Files/CMake/share/cmake-3.19/Modules/FindBoost.cmake:1326 (_Boost_COMPONENT_DEPENDENCIES)
  D:/Program Files/CMake/share/cmake-3.19/Modules/FindBoost.cmake:1935 (_Boost_MISSING_DEPENDENCIES)
  C:/dev/vcpkg/installed/x64-windows/share/boost/vcpkg-cmake-wrapper.cmake:11 (_find_package)
  C:/dev/vcpkg/scripts/buildsystems/vcpkg.cmake:813 (include)
  C:/dev/vcpkg/installed/x64-windows/share/cgal/CGAL_pointmatcher_support.cmake:2 (find_package)
  CMakeLists.txt:96 (include)


CMake Warning at D:/Program Files/CMake/share/cmake-3.19/Modules/FindBoost.cmake:1204 (message):
  New Boost version may have incorrect or missing dependencies and imported
  targets
Call Stack (most recent call first):
  D:/Program Files/CMake/share/cmake-3.19/Modules/FindBoost.cmake:1326 (_Boost_COMPONENT_DEPENDENCIES)
  D:/Program Files/CMake/share/cmake-3.19/Modules/FindBoost.cmake:1935 (_Boost_MISSING_DEPENDENCIES)
  C:/dev/vcpkg/installed/x64-windows/share/boost/vcpkg-cmake-wrapper.cmake:11 (_find_package)
  C:/dev/vcpkg/scripts/buildsystems/vcpkg.cmake:813 (include)
  C:/dev/vcpkg/installed/x64-windows/share/cgal/CGAL_pointmatcher_support.cmake:2 (find_package)
  CMakeLists.txt:96 (include)


CMake Warning at D:/Program Files/CMake/share/cmake-3.19/Modules/FindBoost.cmake:1204 (message):
  New Boost version may have incorrect or missing dependencies and imported
  targets
Call Stack (most recent call first):
  D:/Program Files/CMake/share/cmake-3.19/Modules/FindBoost.cmake:1326 (_Boost_COMPONENT_DEPENDENCIES)
  D:/Program Files/CMake/share/cmake-3.19/Modules/FindBoost.cmake:1935 (_Boost_MISSING_DEPENDENCIES)
  C:/dev/vcpkg/installed/x64-windows/share/boost/vcpkg-cmake-wrapper.cmake:11 (_find_package)
  C:/dev/vcpkg/scripts/buildsystems/vcpkg.cmake:813 (include)
  C:/dev/vcpkg/installed/x64-windows/share/cgal/CGAL_pointmatcher_support.cmake:2 (find_package)
  CMakeLists.txt:96 (include)


CMake Warning at D:/Program Files/CMake/share/cmake-3.19/Modules/FindBoost.cmake:1204 (message):
  New Boost version may have incorrect or missing dependencies and imported
  targets
Call Stack (most recent call first):
  D:/Program Files/CMake/share/cmake-3.19/Modules/FindBoost.cmake:1326 (_Boost_COMPONENT_DEPENDENCIES)
  D:/Program Files/CMake/share/cmake-3.19/Modules/FindBoost.cmake:1935 (_Boost_MISSING_DEPENDENCIES)
  C:/dev/vcpkg/installed/x64-windows/share/boost/vcpkg-cmake-wrapper.cmake:11 (_find_package)
  C:/dev/vcpkg/scripts/buildsystems/vcpkg.cmake:813 (include)
  C:/dev/vcpkg/installed/x64-windows/share/cgal/CGAL_pointmatcher_support.cmake:2 (find_package)
  CMakeLists.txt:96 (include)


CMake Warning at D:/Program Files/CMake/share/cmake-3.19/Modules/FindBoost.cmake:1204 (message):
  New Boost version may have incorrect or missing dependencies and imported
  targets
Call Stack (most recent call first):
  D:/Program Files/CMake/share/cmake-3.19/Modules/FindBoost.cmake:1326 (_Boost_COMPONENT_DEPENDENCIES)
  D:/Program Files/CMake/share/cmake-3.19/Modules/FindBoost.cmake:1935 (_Boost_MISSING_DEPENDENCIES)
  C:/dev/vcpkg/installed/x64-windows/share/boost/vcpkg-cmake-wrapper.cmake:11 (_find_package)
  C:/dev/vcpkg/scripts/buildsystems/vcpkg.cmake:813 (include)
  C:/dev/vcpkg/installed/x64-windows/share/cgal/CGAL_pointmatcher_support.cmake:2 (find_package)
  CMakeLists.txt:96 (include)


CMake Warning at D:/Program Files/CMake/share/cmake-3.19/Modules/FindBoost.cmake:1204 (message):
  New Boost version may have incorrect or missing dependencies and imported
  targets
Call Stack (most recent call first):
  D:/Program Files/CMake/share/cmake-3.19/Modules/FindBoost.cmake:1326 (_Boost_COMPONENT_DEPENDENCIES)
  D:/Program Files/CMake/share/cmake-3.19/Modules/FindBoost.cmake:1935 (_Boost_MISSING_DEPENDENCIES)
  C:/dev/vcpkg/installed/x64-windows/share/boost/vcpkg-cmake-wrapper.cmake:11 (_find_package)
  C:/dev/vcpkg/scripts/buildsystems/vcpkg.cmake:813 (include)
  C:/dev/vcpkg/installed/x64-windows/share/cgal/CGAL_pointmatcher_support.cmake:2 (find_package)
  CMakeLists.txt:96 (include)


Found Boost: C:/dev/vcpkg/installed/x64-windows/include (found version "1.83.0") found components: thread filesystem system program_options date_time chrono atomic 
Configuring done
CMake Error at C:/dev/vcpkg/scripts/buildsystems/vcpkg.cmake:598 (_add_executable):
  Target "registration_with_pointmatcher" links to target "libnabo::nabo" but
  the target was not found.  Perhaps a find_package() call is missing for an
  IMPORTED target, or an ALIAS target is missing?
Call Stack (most recent call first):
  C:/dev/vcpkg/installed/x64-windows/share/cgal/CGAL_CreateSingleSourceCGALProgram.cmake:49 (add_executable)
  CMakeLists.txt:98 (create_single_source_cgal_program)


CMake Error at C:/dev/vcpkg/scripts/buildsystems/vcpkg.cmake:598 (_add_executable):
  Target "registration_with_pointmatcher" links to target "libnabo::nabo" but
  the target was not found.  Perhaps a find_package() call is missing for an
  IMPORTED target, or an ALIAS target is missing?
Call Stack (most recent call first):
  C:/dev/vcpkg/installed/x64-windows/share/cgal/CGAL_CreateSingleSourceCGALProgram.cmake:49 (add_executable)
  CMakeLists.txt:98 (create_single_source_cgal_program)


CMake Error at C:/dev/vcpkg/scripts/buildsystems/vcpkg.cmake:598 (_add_executable):
  Target "registration_with_pointmatcher" links to target "libnabo::nabo" but
  the target was not found.  Perhaps a find_package() call is missing for an
  IMPORTED target, or an ALIAS target is missing?
Call Stack (most recent call first):
  C:/dev/vcpkg/installed/x64-windows/share/cgal/CGAL_CreateSingleSourceCGALProgram.cmake:49 (add_executable)
  CMakeLists.txt:98 (create_single_source_cgal_program)


CMake Error at C:/dev/vcpkg/scripts/buildsystems/vcpkg.cmake:598 (_add_executable):
  Target "registration_with_pointmatcher" links to target "libnabo::nabo" but
  the target was not found.  Perhaps a find_package() call is missing for an
  IMPORTED target, or an ALIAS target is missing?
Call Stack (most recent call first):
  C:/dev/vcpkg/installed/x64-windows/share/cgal/CGAL_CreateSingleSourceCGALProgram.cmake:49 (add_executable)
  CMakeLists.txt:98 (create_single_source_cgal_program)


CMake Error at C:/dev/vcpkg/scripts/buildsystems/vcpkg.cmake:598 (_add_executable):
  Target "registration_with_opengr_pointmatcher_pipeline" links to target
  "libnabo::nabo" but the target was not found.  Perhaps a find_package()
  call is missing for an IMPORTED target, or an ALIAS target is missing?
Call Stack (most recent call first):
  C:/dev/vcpkg/installed/x64-windows/share/cgal/CGAL_CreateSingleSourceCGALProgram.cmake:49 (add_executable)
  CMakeLists.txt:120 (create_single_source_cgal_program)


CMake Error at C:/dev/vcpkg/scripts/buildsystems/vcpkg.cmake:598 (_add_executable):
  Target "registration_with_opengr_pointmatcher_pipeline" links to target
  "libnabo::nabo" but the target was not found.  Perhaps a find_package()
  call is missing for an IMPORTED target, or an ALIAS target is missing?
Call Stack (most recent call first):
  C:/dev/vcpkg/installed/x64-windows/share/cgal/CGAL_CreateSingleSourceCGALProgram.cmake:49 (add_executable)
  CMakeLists.txt:120 (create_single_source_cgal_program)


CMake Error at C:/dev/vcpkg/scripts/buildsystems/vcpkg.cmake:598 (_add_executable):
  Target "registration_with_opengr_pointmatcher_pipeline" links to target
  "libnabo::nabo" but the target was not found.  Perhaps a find_package()
  call is missing for an IMPORTED target, or an ALIAS target is missing?
Call Stack (most recent call first):
  C:/dev/vcpkg/installed/x64-windows/share/cgal/CGAL_CreateSingleSourceCGALProgram.cmake:49 (add_executable)
  CMakeLists.txt:120 (create_single_source_cgal_program)


CMake Error at C:/dev/vcpkg/scripts/buildsystems/vcpkg.cmake:598 (_add_executable):
  Target "registration_with_opengr_pointmatcher_pipeline" links to target
  "libnabo::nabo" but the target was not found.  Perhaps a find_package()
  call is missing for an IMPORTED target, or an ALIAS target is missing?
Call Stack (most recent call first):
  C:/dev/vcpkg/installed/x64-windows/share/cgal/CGAL_CreateSingleSourceCGALProgram.cmake:49 (add_executable)
  CMakeLists.txt:120 (create_single_source_cgal_program)


Generating done”

I'm not sure if the issue lies with the libnabo library not being properly installed, but I have compiled it multiple times following the tutorial at 'https://github.com/norlab-ulaval/libpointmatcher/blob/1.3.1/doc/CompilationWindows.md'. I've been struggling for two days to use libpointmatcher with CGAL, and it's been quite painful.

**Comment by linbeijianbaoxia** [comment]
@lrineau Thank you very much for your suggestion, I will give it a try.
